rust/routee-compass-powertrain/src/routee/energy_traversal_model.rs

```diff
@@ -90,13 +91,58 @@ impl TraversalModel for EnergyTraversalModel {
     fn access_edge(
         &self,
         _v1: &Vertex,
-        _src: &Edge,
+        src: &Edge,
         _v2: &Vertex,
-        _dst: &Edge,
+        dst: &Edge,
         _v3: &Vertex,
-        _state: &TraversalState,
+        state: &TraversalState,
     ) -> Result<Option<TraversalState>, TraversalModelError> {
-        Ok(None)
+        match self.service.headings_table.as_deref() {
+            None => Ok(None),
+            Some(headings_table) => {
+                let src_heading = get_headings(headings_table, src.edge_id)?;
+                let dst_heading = get_headings(headings_table, dst.edge_id)?;
+                let angle = src_heading.next_edge_angle(&dst_heading);
+                let turn = Turn::from_angle(angle)?;
+                let time_cost = match turn {
+                    Turn::NoTurn => {
+                        // no penalty for straight
+                        Time::new(0.0)
+                    }
+                    Turn::SlightRight => {
+                        // 0.5 second penalty for slight right
+                        Time::new(0.5)
+                    }
+                    Turn::Right => {
+                        // 1 second penalty for right
+                        Time::new(1.0)
+                    }
+                    Turn::SharpRight => {
+                        // 1.5 second penalty for sharp right
+                        Time::new(1.5)
+                    }
+                    Turn::SlightLeft => {
+                        // 1 second penalty for slight left
+                        Time::new(1.0)
+                    }
+                    Turn::Left => {
+                        // 2.5 second penalty for left
+                        Time::new(2.5)
+                    }
+                    Turn::SharpLeft => {
+                        // 3.5 second penalty for sharp left
+                        Time::new(3.5)
+                    }
+                    Turn::UTurn => {
+                        // 9.5 second penalty for U-turn
+                        Time::new(9.5)
+                    }
+                };
+                let time = TimeUnit::Seconds.convert(time_cost, &self.service.output_time_unit);
+                let updated_state = add_time_to_state(state, time);
```

Comment:
there should really be an energy update here too. are the Powertrain models well behaved for speed and grade of 0, to give us a kind of "idling energy", or is that something we need to supplant with a configuration value? do we have those values?

Reply:
Right, I think this merits it's own exploration to decide on the best way to represent energy turn costs. It sounds like Jake and Mingdong are planning to take a pass at exploring this.

Also, following up on our discussion earlier today, the powertrain models are not well defined for speeds of 0 and so we would need to provide an explicit "idling energy" if we want to represent the model in those situations. I think we have (or could easily get) the idle costs for each vehicle